Draft: BOPIS Theme App Extension

Banibrata-Manna opened the pull request:
Related to #9

## Context

This adds a theme app extension for BOPIS (Buy Online, Pick Up In Store) so merchants can offer in-store pickup directly from the PDP, without relying on the online store theme's own code.

## What's included

**Pickup Button block** (`pickupButton.liquid`)
- Renders on the product page, either as a **modal** trigger or an **inline** store list, controlled by a block setting.
- Configurable via block settings: button label, search proximity, warehouse inclusion, max stores displayed, "Add Pickup Store property" on the cart line item, weekly store timings visibility, and whether the shopper's home store is included in search results.
- Hides itself if the selected variant isn't available for pickup.

**Store list + pagination** (`snippets/store.liquid`, `snippets/storeListPagination.liquid`)
- Looks up nearby stores via the store search API (`/rest/s1/api/stores`), by zipcode or current location (geolocation), with distance shown per store.
- Paginates results client-side (prev/next), with store timings, phone, and distance shown per store.
- Checks BOPIS inventory per store via `checkBopisInventory` and shows an In Stock / Out of Stock badge; "PICK UP IN STORE" only renders when both pickup is enabled on the block and the store has inventory.

**My Store block + modal** (`myStore.liquid`, `myStoreModal.liquid`)
- Lets a shopper set/change a default pickup store (persisted to `localStorage`, and synced server-side via `setShopifyCustomerDefaultStore`/`getShopifyCustomerDefaultStore` for logged-in customers).
- Search by zipcode or current location, same as the pickup store list.

**Add to cart**
- Adds the selected variant to cart via `/cart/add.js` with the pickup store recorded as a line item property (`_pickupstore`, plus an optional customer-facing address property).
- On success/failure, dispatches `hc:cart:item-added` / `hc:cart:item-add-error` custom events on `document` (with the parsed cart, the added item, and the line item properties in the event detail) so the host theme can hook in its own cart UI (toast, cart icon/count refresh, drawer, etc.) without this extension needing to know that theme's markup.

**Styling & assets**: `hcBopis.css` plus the icon set used across the blocks (location, chevron, clock, phone, check/x).

## Recent fixes on this branch

- Stock badge (In Stock / Out of Stock) was incorrectly tied to the "Enable pick up here" setting, so a merchant using the block as a pure store finder (pickup disabled) would see every store marked "Out of Stock" regardless of actual inventory. Now reflects only actual inventory.
- Fixed pagination Prev/Next button disabled-state bugs (stale state after a variant change, and both buttons getting stuck disabled at pagination boundaries).
- Removed an implicit global variable in the variant-change handler.
- Removed the "Show out of stock stores" block setting (now always effectively on) to keep the initial release simpler; can be reintroduced later if needed.

## Status

Still a draft — open items include verifying end-to-end in a live theme dev session and finishing the multi-page inventory-aware pagination behavior.

Commit:
Pickup Button and Store blocks

Commit:
Used DOM Manipulation to populate stores in Store List Snippet

Commit:
Added support to show Pickup Button based on setting in block and enable pickup for stores in store list

Commit:
Rendering inline and modal store list on basis of pickup button settings

Commit:
Made Similar UI for inline and modal store list and getting stores from oms

Commit:
Added pagination snippet

Commit:
Updated the store snippet to render pagination snippet and hold block setting

Commit:
Updated store search by zip code, added store timings and phone to UI, updated pagination init and reset, refactored the script for reading data properties of element stored in snippets instead passing them as function arguements in script

Commit:
Added max stores display setting in Pickup Button app block, enabled the zipcode searnch from pickup modal

Commit:
Implemented Inventory Lookup and Add to cart

Commit:
fix

Commit:
show out of stock stores and enable warehouse base code

Commit:
Updated App Pickup Button Block to have adjustable width

Commit:
Added product variants in dataset in store list snippet

Commit:
Hide pickup button or inline store list from the PDP if the product is not marked available on theme

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.css:
```diff
@@ -0,0 +1,95 @@
+.store-name {
+    font-size: x-large;
+    font-weight: lighter;
+    padding: 0px 0px 0px 0px;
```
```suggestion
    padding: 0px;
```

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,493 @@
+let PRODUCT_VARIANTS = [];
+
+const PICKUP_TODAY_BTN = document.getElementById('pickup-today-btn');
```
Use uppercase only when having hardcoded and specific variable values for dynamic variable use kebabCase pattern for variable naming.

Banibrata-Manna replied:
Done

Commit:
Moved extension embeds css to file and fixed indentation

Banibrata-Manna committed:
Added custom elements for My Store & it's modal, utility function to fetch all stores

Banibrata-Manna committed:
Moved inline css to asset css file and renamed the element ids and intrduced classnames for reusability

Banibrata-Manna committed:
Added My Store Modal App embed

Banibrata-Manna committed:
Used my store custom element in My Store App Block

Banibrata-Manna committed:
Deleted unused app block and image

Banibrata-Manna committed:
Moving settings to embed block from app block

Banibrata-Manna committed:
Added icons for store contact and store timings

Banibrata-Manna committed:
Changed name of zipcode input id attribute

Banibrata-Manna committed:
Added css rule for store list in my store modal and renamed the modal search class name for reusablity

Banibrata-Manna committed:
Show all stores and my store in my store modal

Banibrata-Manna committed:
Added wrapper over search input and current location search icon

Banibrata-Manna committed:
Added customer ID and Shop's ID in My Store Modal's dataset

Banibrata-Manna committed:
Close Modal

Banibrata-Manna committed:
Minimize Store Name size of Stores in list

Banibrata-Manna committed:
Increased width of Modal

Banibrata-Manna committed:
Hide Set My Store Action from Selected My Store div

Banibrata-Manna committed:
Lookup My Store in Local Storage if customer not logged in

Banibrata-Manna committed:
Code Refactoring

Banibrata-Manna committed:
Show No Store Selected Message if My Store is not Selected by Customer

Banibrata-Manna committed:
Implemented Set Default Store and moved logic to create store list from connected callback to a utility function

Banibrata-Manna committed:
Set id with Store Code value for each store div

Banibrata-Manna committed:
Update My Store in App Block and My Store Modal

Banibrata-Manna committed:
Scroll to top when closing modal and enable page scroll

Banibrata-Manna committed:
Disable page scroll when opening modal

Banibrata-Manna committed:
Clear any default store present in local storage if get default store request doesn't succeed

Banibrata-Manna committed:
Merged Store Name and Timings on My Store App block

Banibrata-Manna committed:
Changed ID for My Store Div Headings

Banibrata-Manna committed:
Avoid Setting selected store here as it is not this function's reponsibility

Banibrata-Manna committed:
Allow Set Store Action if customer does not have default store or on any other store than default store

Banibrata-Manna committed:
Added developer friendly comment

Banibrata-Manna committed:
Added CSS rule to hide element by adding a specific class

Banibrata-Manna committed:
Used CSS escape function to handle special characters and digits in id attibute

Banibrata-Manna committed:
Added support of adjust the visibility of my store in listed stores on search in modal

Banibrata-Manna committed:
Only look for show home store setting when My Store is present in local storage

Banibrata-Manna committed:
Implemented store search by zip, current location and refined code with init and reinit function to refresh store list in modal

Banibrata-Manna committed:
Removed sort by from request body

Banibrata-Manna committed:
Turn current location search off on zipcode search

Banibrata-Manna committed:
Made the Store Contacts div and elements inside it optional and added icons

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.css:
```diff
@@ -0,0 +1,173 @@
+  display: none;
+}
+
+.hc-phone-icon {
```
We can have an icon class that can be applied to all the icons

Banibrata-Manna replied:
used hc-icon

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+
+async function isVariantAvailable(shopifyVariantId) {
+  const variant = await fetchProductVariantById(shopifyVariantId);
+  console.log("*****************************This is the variant availability: ", variant?.available);
```
Remove console logs those were added for debugging purposes.

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+      requestBody.distance = distance;
+    }
+  }
+  if (viewSize) {
+    requestBody.viewSize = viewSize;
+    if (viewIndex) {
+      requestBody.viewIndex = viewIndex;
+    }
+  }
```
```suggestion
  if (viewSize) {
    requestBody.viewSize = viewSize;
  }
  
  if (viewIndex) {
      requestBody.viewIndex = viewIndex;
    }
```

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+
+  requestBody.filters = filters;
+  if (includeWarehouse === 'true') {
+    requestBody.filters.push("storeType: (RETAIL_STORE OR WAREHOUSE OR OUTLET_WAREHOUSE)");
```
```suggestion
    requestBody.filters.push("storeType: (RETAIL_STORE OR WAREHOUSE)");
```

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+
+    const data = await response.json();
+    stores = data?.response?.docs || [];
+    storesFound = data?.response?.numFound || 0;
```
```suggestion
   totalStores = data?.response?.numFound || 0;
```

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+  await searchStoresByLocation(lat, lon);
+}
+
+function addToCart(currentVariantId, quantity = 1, properties) {
```
Move default parameter at the end of parameter list

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      'X-Requested-With': 'XMLHttpRequest'
```
Why is this header required?

Banibrata-Manna replied:
Removed Unnecessary Header

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+  payload.facilityIds = storeCodes;
+  payload.internalNames = [selectedVariantId];
+  payload.productStoreId = 'STORE';
+  payload.inventoryGroupId = 'SHOPIFY_1';
```
```suggestion
  payload.inventoryGroupId = 'FAC_GRP';
```

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+
+async function generateStoreListHTML(container) {
+  if (!container) return;
+  const enablePickup = Boolean(container.dataset.showPickupHere);
```
Do not use Boolean for converting the values as this will return false only for falsy values, for 'false' this will return true.

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+  container.dataset.totalPages = Math.ceil(storesFound / maxStoresToShow);
+  console.log("This is number of total pages: ", container.dataset.totalPages);
+
+  if (storesFound === 0) {
```
```suggestion
  if (!storesFound) {
```

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/assets/hcBopis.js:
```diff
@@ -0,0 +1,1036 @@
+  : '';
+
+  stores.forEach(store => {
+    const inStock = storesWithInventory?.includes(store.storeCode);
```
```suggestion
    const isInStock = storesWithInventory?.includes(store.storeCode);
```

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/snippets/storeListPagination.liquid:
```diff
@@ -0,0 +1,28 @@
+<div id="pagination" style="display:flex; justify-content:center; gap:1rem; margin-top:1rem; align-items:center;">
```
Remove inline styling and move it to specific file to add style block at the end of the file.

Banibrata-Manna replied:
Done

ymaheshwari1 commented on extensions/hc-dev-theme-extension/snippets/storeListPagination.liquid:
```diff
@@ -0,0 +1,28 @@
+<div id="pagination" style="display:flex; justify-content:center; gap:1rem; margin-top:1rem; align-items:center;">
+  <button id="prev-page" disabled class="nav-btn">&lt;&lt;</button>
```
Use icons in place of &lt; and &gt;

Banibrata-Manna replied:
Used Images from Polaris, instead of arrows

Banibrata-Manna committed:
Used common class css rules for phone and clock icons

Banibrata-Manna committed:
Keep View Index Independent and remove outlet warehouse from store lookup

Banibrata-Manna committed:
Function to create pickup store divs

Banibrata-Manna committed:
Disable Page Scroll on Opening  pickup modal

Banibrata-Manna committed:
Renamed variable

Banibrata-Manna committed:
Renamed flag variable

Banibrata-Manna committed:
Moved default variable to last in function signature and removed unnecessary header

Banibrata-Manna committed:
Added wrapper over store divs changed pickup-btn id to class and used span in store details and flex

Banibrata-Manna committed:
Simplified padding rule

Banibrata-Manna committed:
Wrap liquid objects in quotes

Banibrata-Manna committed:
Removed unused css on id pickup-btn

Banibrata-Manna committed:
Moved Pagination Snippets CSS to css file and used image icons instead of arrows

Banibrata-Manna committed:
Added Icons

Banibrata-Manna committed:
Support My Store Visibility Setting in Pickup Store List

Banibrata-Manna committed:
Used Icon for product stock info

Banibrata-Manna committed:
Updated size of icons and custom line margin

Banibrata-Manna committed:
Return for empty zipcode input

Banibrata-Manna committed:
Changed Location Icons and updated current location search event listener

Banibrata-Manna committed:
Fix store view size exceeding per-page limit on simultaneous pagination

Banibrata-Manna committed:
Added default display style pagination controls

Banibrata-Manna committed:
My Store skeleton div

Banibrata-Manna committed:
Different CSS rules for Pickup Button and My Store info in Modal and inline moda

Banibrata-Manna committed:
Added necessary setting to use in the script

Banibrata-Manna committed:
Utility function to create My Store Pickup info and action element in inline mode

Banibrata-Manna committed:
Show My Store Info and Action in Pickup Modal and inline store list

Banibrata-Manna committed:
Renamed element class and updated my store align rule

Banibrata-Manna committed:
Added Settings in data attibutes to pass them to script

Banibrata-Manna committed:
Added a skeleton div for My Store Inline mode and renamed the setting in data attribute

Banibrata-Manna committed:
Implemented the Update of My Store in Pickup Block

Banibrata-Manna committed:
Moved Pickup Settings from dataset to script variables and updated add to cart action

Banibrata-Manna committed:
Util function to create pickup buttons for stores

Banibrata-Manna committed:
Create Pickup In Store Buttons by util function

Banibrata-Manna committed:
Show and Hide My Store Pickup Button based on inventory

Banibrata-Manna committed:
Added missing store paramater in create pickup button function

Banibrata-Manna committed:
Added util function to get my store from local storage

Banibrata-Manna committed:
Added default value in Pickup Stores Prosimity on search

Banibrata-Manna committed:
Added default value for store search proximity

Banibrata-Manna committed:
changed case for product variants variable

Banibrata-Manna committed:
Changed Default Property of Pickup Item Propoerty Label

ymaheshwari1 commented on extensions/hc-dev-theme-extension/blocks/pickupModal.liquid:
```diff
@@ -0,0 +1,72 @@
+      } catch (error) {
+        console.error("Error in Searching stores by current location");
+        // reset the location icons state to what it was.
+        if (pickLocation) {
```
pickLocation is defined only in the try block, and thus will throw an error here of undefined

Banibrata-Manna replied:
Logging only error message in catch block.

Banibrata-Manna committed:
Removed unwanted code

Banibrata-Manna committed:
Show distance if present in store details

Banibrata-Manna committed:
Reset current location icon data on zipcode search

Banibrata-Manna committed:
Updated zipcode search in Pickup and My Store Modals to form

Banibrata-Manna committed:
Used hard coded css rules instead of configurable styling

Banibrata-Manna committed:
Removed dataset value parameters in show pickup modal and show store proximity radius in input placeholder of zipcode search

Banibrata-Manna committed:
Removed Unused settings

Banibrata-Manna committed:
Fixed unwanted inventory lookup with undefined as store id when My Store is selected

Banibrata-Manna committed:
Show My Store Heading only in Modal

Banibrata-Manna committed:
Removed Hard Coded customer ID

Banibrata-Manna committed:
Removed width property from Pickup Today Button Styling

Banibrata-Manna committed:
Show Store Search radius in My Store Modal zipcode search placeholder

Banibrata-Manna committed:
Pass & set Setting value to adjust store timings visibility

Banibrata-Manna committed:
Remove old event listeners when initializing pagination

Banibrata-Manna committed:
Used util function to get My Store

Banibrata-Manna committed:
Listen to variant selection event only by class name and added todo

Banibrata-Manna committed:
Changed: updated response handling of store lookup api

Banibrata-Manna committed:
Fixed: build failure due to incorrect agument in render snippet

Banibrata-Manna committed:
Changed: Removed show out stock of stores checkbox from theme editor and keep it true by default

Banibrata-Manna committed:
Added: dispatch custom pickup events for themes to identify pickup items are being added to cart, best point to do general things in themes like update cart item count or open mini cart drawer etc.

Banibrata-Manna committed:
Changed: show inventory stock status irrespective of the pickup is enabled
make inventory store list constant

Banibrata-Manna committed:
Changed: Print error when looking up stores by shopper's current location

Banibrata-Manna committed:
Changed: refactor pickup button/modal/store list into custom elements

Convert the pickup button, pickup modal, and store list from
plain functions over cached DOM refs into guarded custom elements
(<hc-pickup-button>, <hc-pickup-modal>, <hc-pickup-store-list>),
matching the existing my-store/my-store-modal pattern. Based on
conventions studied in Shopify's Dawn and Horizon theme source.

- Native <dialog> for the pickup modal (.showModal()/.close()),
  replacing manual style.display toggling
- PickupConfig/ProductVariantState singletons replace the old
  pickupBlockSettings/storeListProperties/selectedProductVariant
  globals, preserving the button-wins-over-modal settings semantics
- Two new internal events (hc:pickup:variant-changed,
  hc:pickup:store-changed) replace direct cross-element
  document.querySelector reach-arounds between My Store and the
  inline pickup list
- alert() replaced with an inline, non-blocking error message
- AbortController cancellation on store/geocode search requests
- Removed dead code (pickupStores, getZipCode stub,
  myStoreBlockSettings)

patelanil reviewed (commented):
**Log message format review**

I reviewed the log/error messages added in this PR against our best practice: [log-messages.md](https://github.com/saastechacademy/foundation/blob/main/moqui-framework/bestpractices/log-messages.md).

These added messages do not follow the `[Entity] [Context] - [Action/Outcome/Issue]` pattern:

1. `Error parsing product variants dataset:`
   Suggest: `Product Variants - Failed to parse dataset`
2. `Error getting location:`
   Suggest: `Location - Failed to get current location`
3. `Error fetching stores:`
   Suggest: `Stores - Failed to fetch`
4. `Error fetching latitude and longitude:`
   Suggest: `Location - Failed to fetch latitude and longitude`
5. `addToCart called without a selected variant — aborted.`
   Suggest: `Cart - Add to cart aborted: no variant selected`
6. `Error in Searching stores by current location:`
   Suggest: `Stores - Failed to search by current location`
7. `Empty Input`
   Suggest: `Store Search - Aborted: empty input`
8. `Error fetching customer default store:`
   Suggest: `Customer Default Store - Failed to fetch`

...and 6 more in the same pattern.

Dynamic IDs belong in the `[Context]` block and the action text should stay constant, so messages group and count cleanly in Grafana/Loki during log analysis. Please align these before merge.
